dev/src/bundle.ts

```diff
@@ -0,0 +1,184 @@
+    } else if (querySnapshot instanceof QuerySnapshot) {
+      this.addNamedQuery(documentOrName, querySnapshot);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    if (this.namedQueries.has(name)) {
+      throw new Error(`Query name conflict: ${name} is already added.`);
+    }
+
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private lengthPrefixedBuffer(payload: string): Buffer {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+    return Buffer.concat([lengthBuffer, buffer]);
+  }
+
+  stream(): NodeJS.ReadableStream {
+    const readable = new Readable({
+      objectMode: false,
+      read: () => {
+        const metadata: firestore.IBundleMetadata = {
+          id: this.bundleId,
+          createTime: this.latestReadTime.toProto().timestampValue,
+        };
+        readable.push(
+          this.lengthPrefixedBuffer(
+            JSON.stringify({metadata} as firestore.IBundleElement)
+          )
+        );
+
+        for (const namedQuery of this.namedQueries.values()) {
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({namedQuery} as firestore.IBundleElement)
+            )
+          );
+        }
+
+        for (const bundledDocument of this.documents.values()) {
+          const documentMetadata: firestore.IBundledDocumentMetadata =
+            bundledDocument.metadata;
+          const document: api.IDocument = bundledDocument.document;
+
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({documentMetadata} as firestore.IBundleElement)
+            )
+          );
+          readable.push(
+            this.lengthPrefixedBuffer(
+              JSON.stringify({document} as firestore.IBundleElement)
+            )
+          );
+        }
+        readable.push(null);
+      },
+    });
+
+    return readable;
+  }
+
+  build(): Promise<Buffer> {
```

Comment:
This could be simplified:

```
 async build(): Promise<Buffer> {
    const stream = this.stream();
    let data = Buffer.alloc(0);
    for await (const chunk of stream) {
      data = Buffer.concat([data, chunk as Uint8Array]);
    }
    return data;
  }
```

Reply:
`async` is not part of the API, only `Promise<Buffer>` is. There are some subtle differences for `async` functions but I don't think the differences apply here.

```
let i = 0;
function inc() : Promise.resolve() {
  ++i
  return Promise.resolve();
}
inc() // `i` is now 1
```
Versus:
```
let i = 0;
async function inc() : Promise.resolve() {
  ++i
  return Promise.resolve();
}
inc() // `i` is now 0, since the entire function is executed via the Promise chain
await Promise.resolve(); // `i` is now 1
```